src/main/scala/is/hail/utils/HashMethods.scala

```diff
@@ -0,0 +1,58 @@
+package is.hail.utils
+
+import java.util.Random
+
+object SimpleTabulationHash32 {
+
+  //TODO: replace java.util.Random with a better PRNG
+  def apply(seed: Long): SimpleTabulationHash32 = {
+    val random = new Random(seed)
+    new SimpleTabulationHash32(Array.fill[Array[Int]](4)(random.ints(256).toArray()))
+  }
+}
+
+// see e.g. Thorup, "Fast and Powerful Hashing using Tabulation", for explanation of both simple and twisted tabulation
+// hashing
+class SimpleTabulationHash32(table: Array[Array[Int]]) extends (Int => Int) {
+  require(table.length == 4)
+  require(table(0).length == 256)
+
+  override def apply(key: Int): Int = {
+    var out: Int = 0
+    var keyBuffer: Int = key
+    var i = 0
+    while (i < 4) {
+      out ^= table(i)(keyBuffer & 255)
+      i += 1
+      keyBuffer >>>= 8
+    }
+    out
+  }
+}
+
+object TwistedTabulationHash32 {
+
+  //TODO: replace java.util.Random with a better PRNG
+  def apply(seed: Long): TwistedTabulationHash32 = {
+    val random = new Random(seed)
+    new TwistedTabulationHash32(Array.fill[Array[Long]](4)(random.longs(256).toArray()))
+  }
+}
+
+class TwistedTabulationHash32(table: Array[Array[Long]]) extends (Int => Int) {
+  require(table.length == 4)
+  for (i <- 0 to 3) require(table(i).length == 256)
```

Comment:
We might in the future want to run disabling asserts/requires, etc.  Because of that, I prefer to keep all assert-related computation inside the assert.  Therefore, I'd write this as `require(table.forall(_.length == 256))`.